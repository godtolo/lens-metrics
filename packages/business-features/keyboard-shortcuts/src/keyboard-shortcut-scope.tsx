import React from "react";

export interface KeyboardShortcutScopeProps {
  id: string;
  children: React.ReactNode;
}

export const KeyboardShortcutScope = ({ id, children }: KeyboardShortcutScopeProps) => (
  <div data-keyboard-shortcut-scope={id} data-keyboard-shortcut-scope-test={id} tabIndex={-1}>
    {children}
  </div>
);
