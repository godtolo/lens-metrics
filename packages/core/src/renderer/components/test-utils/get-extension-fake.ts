/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { LensMainExtensionParts } from "../../../extensions/lens-main-extension";
import { LensMainExtension } from "../../../extensions/lens-main-extension";
import type { LensRendererExtensionParts } from "../../../extensions/lens-renderer-extension";
import { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import type { SetRequired } from "type-fest";
import type { LensExtensionManifest } from "../../../features/extensions/common/installed-extension";

export class TestExtensionMain extends LensMainExtension {}
export class TestExtensionRenderer extends LensRendererExtension {}

type FakeManifest = SetRequired<Partial<LensExtensionManifest>, "name">;

export interface FakeExtensionOptions {
  id: string;
  manifest: FakeManifest;
  rendererOptions?: Partial<LensRendererExtensionParts>;
  mainOptions?: Partial<LensMainExtensionParts>;
}

export const getExtensionFakeForMain = ({ id, manifest, mainOptions: options = {}}: FakeExtensionOptions) => (
  Object.assign(
    new TestExtensionMain({
      id,
      absolutePath: "irrelevant",
      isBundled: false,
      isCompatible: false,
      isEnabled: false,
      manifest: {
        version: "1.0.0",
        engines: {
          lens: "^5.5.0",
        },
        ...manifest,
      },
      manifestPath: "irrelevant",
    }),
    options,
  )
);

export const getExtensionFakeForRenderer = ({ id, manifest, rendererOptions: options = {}}: FakeExtensionOptions) => (
  Object.assign(
    new TestExtensionRenderer({
      id,
      absolutePath: "irrelevant",
      isBundled: false,
      isCompatible: false,
      isEnabled: false,
      manifest: {
        version: "1.0.0",
        engines: {
          lens: "^5.5.0",
        },
        ...manifest,
      },
      manifestPath: "irrelevant",
    }),
    options,
  )
);
