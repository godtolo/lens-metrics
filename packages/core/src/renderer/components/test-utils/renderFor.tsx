/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import React from "react";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import type { DiContainer } from "@ogre-tools/injectable";
import { DiContextProvider } from "@ogre-tools/injectable-react";

export type DiRender = (ui: React.ReactElement) => RenderResult;

type DiRenderFor = (di: DiContainer) => DiRender;

const getDOM = (di: DiContainer, ui: React.ReactElement) => (
  <DiContextProvider value={{ di }}>
    {ui}
  </DiContextProvider>
);

export const renderFor: DiRenderFor = (di) => (ui) => {
  const result = render(getDOM(di, ui));

  return {
    ...result,
    rerender: (ui) => result.rerender(getDOM(di, ui)),
  };
};
