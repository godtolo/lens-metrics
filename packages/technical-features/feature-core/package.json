{
  "name": "@k8slens/feature-core",
  "private": false,
  "version": "0.0.1",
  "description": "Code that is common to all Features and those registering them.",
  "type": "commonjs",
  "files": [
    "dist"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/lensapp/lens.git"
  },
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "author": {
    "name": "OpenLens Authors",
    "email": "info@k8slens.dev"
  },
  "license": "MIT",
  "homepage": "https://github.com/lensapp/lens",
  "scripts": {
    "build": "webpack",
    "dev": "webpack --mode=development --watch",
    "test": "jest --coverage --runInBand"
  },
  "peerDependencies": {
    "@ogre-tools/injectable": "^15.1.1"
  }
}
