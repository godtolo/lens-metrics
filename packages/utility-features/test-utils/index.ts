export * from "./src/flush-promises";
export * from "./src/get-global-override-for-function";
export * from "./src/get-global-override";
export * from "./src/get-promise-status";
