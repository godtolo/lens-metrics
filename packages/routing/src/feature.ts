import { getFeature } from "@k8slens/feature-core";
import { autoRegister } from "@ogre-tools/injectable-extension-for-auto-registration";
import { reactApplicationFeature } from "@k8slens/react-application";

export const routingFeature = getFeature({
  id: "routing",

  register: (di) => {
    autoRegister({
      di,
      targetModule: module,
      getRequireContexts: () => [
        require.context("./", true, /\.injectable\.(ts|tsx)$/),
      ],
    });
  },

  dependencies: [reactApplicationFeature],
});
