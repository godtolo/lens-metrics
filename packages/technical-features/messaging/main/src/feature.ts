import { autoRegister } from "@ogre-tools/injectable-extension-for-auto-registration";
import { getFeature } from "@k8slens/feature-core";

export const messagingFeatureForMain = getFeature({
  id: "messaging-for-main",

  register: (di) => {
    autoRegister({
      di,
      targetModule: module,

      getRequireContexts: () => [require.context("./", true, /\.injectable\.(ts|tsx)$/)],
    });
  },
});
