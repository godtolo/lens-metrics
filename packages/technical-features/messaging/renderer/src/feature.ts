import { autoRegister } from "@ogre-tools/injectable-extension-for-auto-registration";
import { getFeature } from "@k8slens/feature-core";

export const messagingFeatureForRenderer = getFeature({
  id: "messaging-for-renderer",

  register: (di) => {
    autoRegister({
      di,
      targetModule: module,

      getRequireContexts: () => [require.context("./", true, /\.injectable\.(ts|tsx)$/)],
    });
  },
});
