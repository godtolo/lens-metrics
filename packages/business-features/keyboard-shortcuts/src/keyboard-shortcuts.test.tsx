import userEvent from "@testing-library/user-event";
import type { RenderResult } from "@testing-library/react";
import { createContainer, DiContainer, getInjectable } from "@ogre-tools/injectable";
import { registerInjectableReact } from "@ogre-tools/injectable-react";
import { keyboardShortcutInjectionToken } from "./keyboard-shortcut-injection-token";
import { registerFeature } from "@k8slens/feature-core";
import { keyboardShortcutsFeature } from "./feature";
import { renderFor } from "@k8slens/test-utils";
import React from "react";
import { KeyboardShortcutScope } from "./keyboard-shortcut-scope";
import { Discover, discoverFor } from "@k8slens/react-testing-library-discovery";
import { KeyboardShortcutListener } from "./keyboard-shortcut-listener";

const TestComponent = () => (
  <KeyboardShortcutListener>
    <div>
      <div>
        <KeyboardShortcutScope id="some-scope">
          <div />
        </KeyboardShortcutScope>
      </div>
    </div>
  </KeyboardShortcutListener>
);

describe("keyboard-shortcuts", () => {
  let di: DiContainer;
  let invokeMock: jest.Mock;

  beforeEach(() => {
    di = createContainer("irrelevant");

    registerInjectableReact(di);

    registerFeature(di, keyboardShortcutsFeature);

    invokeMock = jest.fn();

    const someKeyboardShortcutInjectable = getInjectable({
      id: "some-keyboard-shortcut",

      instantiate: () => ({
        binding: "Escape",
        invoke: () => invokeMock("esc-in-root"),
      }),

      injectionToken: keyboardShortcutInjectionToken,
    });

    const someScopedKeyboardShortcutInjectable = getInjectable({
      id: "some-scoped-keyboard-shortcut",

      instantiate: () => ({
        binding: "Escape",
        invoke: () => invokeMock("esc-in-scope"),
        scope: "some-scope",
      }),

      injectionToken: keyboardShortcutInjectionToken,
    });

    const someOtherKeyboardShortcutInjectable = getInjectable({
      id: "some-other-keyboard-shortcut",

      instantiate: () => ({
        binding: "something-else-than-esc",
        invoke: () => invokeMock("something-else-than-esc"),
      }),

      injectionToken: keyboardShortcutInjectionToken,
    });

    di.register(
      someKeyboardShortcutInjectable,
      someScopedKeyboardShortcutInjectable,
      someOtherKeyboardShortcutInjectable,
    );
  });

  describe("when rendered", () => {
    let rendered: RenderResult;
    let discover: Discover;

    beforeEach(async () => {
      const render = renderFor(di);

      rendered = render(<TestComponent />);

      discover = discoverFor(() => rendered);
    });

    it("renders", () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });

    it("given focus is in the body, when pressing the shortcut, calls shortcut in global scope", () => {
      userEvent.keyboard("{Escape}");

      expect(invokeMock.mock.calls).toEqual([["esc-in-root"]]);
    });

    it("given focus inside a nested scope, when pressing the shortcut, calls only the callback for the scope", () => {
      const result = discover.getSingleElement("keyboard-shortcut-scope", "some-scope");

      const discoveredHtml = result.discovered as HTMLDivElement;

      discoveredHtml.focus();

      userEvent.keyboard("{Escape}");

      expect(invokeMock.mock.calls).toEqual([["esc-in-scope"]]);
    });

    it("given conflicting shortcut, when pressing the shortcut, calls both callbacks", () => {
      const conflictingShortcutInjectable = getInjectable({
        id: "some-conflicting-keyboard-shortcut",

        instantiate: () => ({
          binding: "Escape",
          invoke: () => invokeMock("conflicting-esc-in-root"),
        }),

        injectionToken: keyboardShortcutInjectionToken,
      });

      di.register(conflictingShortcutInjectable);

      userEvent.keyboard("{Escape}");

      expect(invokeMock.mock.calls).toEqual([["esc-in-root"], ["conflicting-esc-in-root"]]);
    });

    [
      {
        scenario: "given shortcut without modifiers, when shortcut is pressed, calls the callback",
        binding: { code: "Escape" },
        keyboard: "{Escape}",
        shouldCallCallback: true,
      },
      {
        scenario:
          "given shortcut without modifiers, when shortcut is pressed but with modifier, does not call the callback",
        binding: { code: "F1" },
        keyboard: "{Meta>}[F1]",
        shouldCallCallback: false,
      },
      {
        scenario: "given shortcut with meta modifier, when shortcut is pressed, calls the callback",

        binding: { meta: true, code: "F1" },
        keyboard: "{Meta>}[F1]",
        shouldCallCallback: true,
      },
      {
        scenario:
          "given shortcut with shift modifier, when shortcut is pressed, calls the callback",

        binding: { shift: true, code: "F1" },
        keyboard: "{Shift>}[F1]",
        shouldCallCallback: true,
      },
      {
        scenario: "given shortcut with alt modifier, when shortcut is pressed, calls the callback",
        binding: { altOrOption: true, code: "F1" },
        keyboard: "{Alt>}[F1]",
        shouldCallCallback: true,
      },
      {
        scenario: "given shortcut with ctrl modifier, when shortcut is pressed, calls the callback",
        binding: { ctrl: true, code: "F1" },
        keyboard: "{Control>}[F1]",
        shouldCallCallback: true,
      },
      {
        scenario: "given shortcut with all modifiers, when shortcut is pressed, calls the callback",

        binding: { ctrl: true, altOrOption: true, shift: true, meta: true, code: "F1" },
        keyboard: "{Meta>}{Shift>}{Alt>}{Control>}[F1]",
        shouldCallCallback: true,
      },
    ].forEach(({ binding, keyboard, scenario, shouldCallCallback }) => {
      // eslint-disable-next-line jest/valid-title
      it(scenario, () => {
        const shortcutInjectable = getInjectable({
          id: "shortcut",

          instantiate: () => ({
            binding,
            invoke: invokeMock,
          }),

          injectionToken: keyboardShortcutInjectionToken,
        });

        di.register(shortcutInjectable);

        userEvent.keyboard(keyboard);

        if (shouldCallCallback) {
          // eslint-disable-next-line jest/no-conditional-expect
          expect(invokeMock).toHaveBeenCalled();
        } else {
          // eslint-disable-next-line jest/no-conditional-expect
          expect(invokeMock).not.toHaveBeenCalled();
        }
      });
    });
  });
});
