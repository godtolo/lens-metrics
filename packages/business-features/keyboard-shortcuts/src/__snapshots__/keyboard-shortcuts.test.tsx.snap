// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keyboard-shortcuts when rendered renders 1`] = `
<body>
  <div>
    <div>
      <div>
        <div
          data-keyboard-shortcut-scope="some-scope"
          data-keyboard-shortcut-scope-test="some-scope"
          tabindex="-1"
        >
          <div />
        </div>
      </div>
    </div>
  </div>
</body>
`;
